### ISSUE TITLE
Support hexagonal grids with GridMap

### ISSUE THREAD

**Issue by octetdev2** [godotengine/godot-proposals#4337]
### Describe the project you are working on

Godot engine improvements.

### Describe the problem or limitation you are having in your project

There is no built-in support for hexagonal tiling.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Support hexagon grids for the `GridMap` node.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

We'll add cell shape, layout and and offset enums to support non square grids the same way TileMap does.

## Add a cell shape enum:

```cpp
enum CellShape {
  CELL_SHAPE_SQUARE,
  CELL_SHAPE_HEXAGON,
};

void set_cell_shape(CellShape p_shape);
CellShape get_cell_shape() const;
```

## Add a cell layout enum:

```cpp
enum CellLayout {
  CELL_LAYOUT_STACKED,
  CELL_LAYOUT_STACKED_OFFSET,
  CELL_LAYOUT_STAIRS_RIGHT,
  CELL_LAYOUT_STAIRS_DOWN,
  CELL_LAYOUT_DIAMOND_RIGHT,
  CELL_LAYOUT_DIAMOND_DOWN,
};

void set_cell_layout(CellLayout p_layout);
CellLayout get_cell_layout() const;
```

## Add an offset axis enum:

```cpp
enum CellOffsetAxis {
  CELL_SHAPE_SQUARE,
  CELL_SHAPE_HEXAGON,
};

void set_cell_offset_axis(CellOffsetAxis p_offset_axis);
CellOffsetAxis get_cell_offset_axis() const;
```

Semantics will pretty much be the same as TileMap except that we do it in the xz plane.
There is a bit of duplication here, that is intentional. My intent is to reuse code where applicable when GridMap becomes TileMap3D.

## PoC
https://user-images.githubusercontent.com/102470155/161424657-47ec247d-1b84-40db-93c1-f59a8bef7094.mov

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This is a design limitation within the GridMap node.

### Is there a reason why this should be core and not an add-on in the asset library?

This is about improving GridMap functionality out of the box.

**Comment by Wasabi-Cheetah** [comment]
I was about to request this as well, this would come in VERY handy. If it is possible, it would be a really awesome thing if it could wrap around spheres to make tileable planets.

**Comment by Wasabi-Cheetah** [comment]
Oh, thank you!

**Comment by octetdev2** [comment]
Shaders only get you so far tho.

I'd be for having a projection option for determining what is the tiling plane.

**Comment by Wasabi-Cheetah** [comment]
yeah, I've learned a bit of what I was going for and as per my research into this, a curvature shader won't cut it, since I need to tile the mesh, fortunately I found this Stack Exchange post. https://stackoverflow.com/questions/46777626/mathematically-producing-sphere-shaped-hexagonal-grid Hope it helps others as well.

**Comment by clayjohn** [comment]
We discussed today in the proposals review meeting. In general we are supportive of adding a hexagon mode, but we need more information about the use-case for the alternative cell layouts and offset axis. Right now it is unclear whether all these options are useful for users or are even needed. Our primary concern is that they add significant complexity to the implementation, but may not be used by many users. 

So, to move forward could someone clearly lay out the uses for the additional cell layouts and offset axis? The benefit of hexagon mode is much more obvious by contrast.

**Comment by kukelekuuk** [comment]
The 2D Tilemap implementation offers a half-offset option. Which I believe is meant to be used for hexagonal tilemaps. I'm not sure what other use-case it has.

Tilemap has square (with half offset for hexagon) and isometric. Maybe both 2D and 3D tilemaps should have square, triangle and hexagon modes by default. (and add an isometric option under each in 2D?) 
Then a "custom" mode can have all the complex settings for the advanced user. Where instead of half-offset it just has a normal offset slider (for each axis)

That seems like the more user-friendly option without sacrificing functionality for the advanced users.

**Comment by alazifk** [comment]
Please consider combining hexagon and triangle and square and triangle.

![Untitled](https://user-images.githubusercontent.com/93336644/191058291-91c0da1c-4cf8-4849-bc57-b827623e3a0e.jpg)

**Comment by goosewoman** [comment]
Has there been any word on any progress on this front?

**Comment by AThousandShips** [comment]
> But I think feature parity between tilemaps and gridmaps should be obvious.

Only if it's useful, if people actually need it, there's a signficant difference between those two types, they are not the same and their uses are not the same

**Comment by BSChad** [comment]
Since octetdev seems inactive, I revived (rebased and cleaned up) https://github.com/godotengine/godot/pull/59842 because I am about to make a game that uses hex grids, and I'd love to be able to use the GridMap for this.

Behold, a GridMap based hex grid regenerating from the MeshLibrary on key press.

https://github.com/godotengine/godot-proposals/assets/167169/271f53a3-22ec-46f2-b3d3-ef771d030f1c

**Comment by arjhun** [comment]
wow @BSChad I was waiting for a long time on this one!! Build myself a little game without grid map (all in code) But this will be a live saver, also for people wanting to build boardgame style 3d games etc.

**Comment by Malkverbena** [comment]
Any update about this feature?

**Comment by m21-cerutti** [comment]
@Malkverbena I have seen a PR for that, https://github.com/godotengine/godot/pull/85890

**Comment by sean2077** [comment]
If people want to make a game like [Romance of the Three Kingdoms XI with Power Up Kit](https://store.steampowered.com/app/628070/Romance_of_the_Three_Kingdoms_XI_with_Power_Up_Kit/), they will need half-offset square or hexagon Gridmap. (the following is RTX X1 gameplay scene)

<img width="1924" height="1119" alt="Image" src="https://github.com/user-attachments/assets/7af23394-145c-470d-8aed-616b2dab94f7" />
